File: worker/src/db/documents.ts
import type { Env } from "../types";
import { utcNow } from "./opportunities";

export type DocumentRow = {
  id: number;
  user_id: number;
  name: string;
  doc_type: string;
  size_bytes: number | null;
  status: string | null;
  storage_ref: string | null;
  content_type: string | null;
  created_at: string;
  updated_at: string;
};

export async function listDocuments(env: Env, userId: number): Promise<DocumentRow[]> {
  const { results } = await env.DB.prepare("SELECT * FROM documents WHERE user_id = ? ORDER BY created_at DESC")
    .bind(userId)
    .all<DocumentRow>();
  return results;
}

export async function getDocument(env: Env, userId: number, documentId: number): Promise<DocumentRow | null> {
  const row = await env.DB.prepare("SELECT * FROM documents WHERE user_id = ? AND id = ?")
    .bind(userId, documentId)
    .first<DocumentRow>();
  return row ?? null;
}

export async function createDocument(
  env: Env,
  userId: number,
  fields: {
    name: string;
    docType: string;
    sizeBytes: number | null;
    status: string | null;
    storageRef: string | null;
    contentType: string | null;
  }
): Promise<DocumentRow> {
  const now = utcNow();
  const result = await env.DB.prepare(
    `INSERT INTO documents (user_id, name, doc_type, size_bytes, status, storage_ref, content_type, created_at, updated_at)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`
  )
    .bind(userId, fields.name, fields.docType, fields.sizeBytes, fields.status, fields.storageRef, fields.contentType, now, now)
    .run();
  const row = await env.DB.prepare("SELECT * FROM documents WHERE id = ?")
    .bind(result.meta.last_row_id)
    .first<DocumentRow>();
  if (!row) throw new Error("document insert did not persist");
  return row;
}

export async function deleteDocument(env: Env, userId: number, documentId: number): Promise<DocumentRow | null> {
  const row = await getDocument(env, userId, documentId);
  if (!row) return null;
  await env.DB.prepare("DELETE FROM documents WHERE user_id = ? AND id = ?").bind(userId, documentId).run();
  return row;
}

File: models/documents.py
"""Document metadata only -- actual file bytes stay client-side (IndexedDB) for now.
See the pre-production plan's Phase 2 note: server-side file storage (e.g. an
S3-compatible bucket) is separate, deferred scope."""
from __future__ import annotations

from models.db import db_connect, db_lock
from models.opportunities import utc_now


def list_documents(user_id: int) -> list[dict]:
    with db_lock, db_connect() as connection:
        rows = connection.execute("SELECT * FROM documents WHERE user_id = ? ORDER BY created_at DESC", (user_id,)).fetchall()
    return [dict(row) for row in rows]


def create_document(user_id: int, name: str, doc_type: str, size_bytes: int | None, status: str | None) -> dict:
    now = utc_now()
    with db_lock, db_connect() as connection:
        cursor = connection.execute(
            "INSERT INTO documents (user_id, name, doc_type, size_bytes, status, created_at, updated_at) VALUES (?, ?, ?, ?, ?, ?, ?)",
            (user_id, name, doc_type, size_bytes, status, now, now),
        )
        row = connection.execute("SELECT * FROM documents WHERE id = ?", (cursor.lastrowid,)).fetchone()
    return dict(row)


def delete_document(user_id: int, document_id: int) -> None:
    with db_lock, db_connect() as connection:
        connection.execute("DELETE FROM documents WHERE user_id = ? AND id = ?", (user_id, document_id))
